include/gc/ExecutionEngine/Driver/Driver.h

```diff
@@ -0,0 +1,69 @@
+//===-- Driver.h - The top-level MLIR compiler driver -----------*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef GC_EXECUTIONENGINE_DRIVER_DRIVER_H
+#define GC_EXECUTIONENGINE_DRIVER_DRIVER_H
+
+#include "mlir/ExecutionEngine/CRunnerUtils.h"
+#include "mlir/ExecutionEngine/ExecutionEngine.h"
+#include <memory>
+#include <string_view>
+
+namespace mlir {
+class DialectRegistry;
+namespace gc {
+
+const DialectRegistry &initCompilerAndGetDialects();
+
+// the pointers to XXXMemRefType
+using GeneralMemrefPtr = void *;
+using JitModuleFuncT = void (*)(void **);
+
+struct DriverOptions {
+  // the optimization level for the LLVM-JIT
+  llvm::CodeGenOptLevel jitCodeGenOptLevel = llvm::CodeGenOptLevel::Aggressive;
+  // whether to run the MLIR transformation passes
+  bool runTransforms = true;
+  // todo: target machine, etc.
+};
+
+class JitModule {
+public:
+  static llvm::Expected<std::shared_ptr<JitModule>>
+  create(Operation *op, const DriverOptions &options = {});
+
+  // args should be an array of XXXMemrefType*
+  void call(GeneralMemrefPtr *args, std::size_t numArgs) {
+    // Silly code, MLIR execution engine requires pointers of real args as
+    // inputs
+    llvm::SmallVector<void *, 32> realargs;
```

Comment:
Better to cache this vector.

Reply:
We can't cache it. The args may change and we need to consider multi-threading.

Reply:
Just cache the vector to avoid creating it many times. We still need to set the pointer into it for each execution.

Reply:
Let's optimize this part later when it becomes a performance bottleneck.

Reply:
What if multiple threads are calling `JitWrapper::call`? `llvm::SmallVector<void *, 32> realargs ` allocates on stack in the happy path, so usually we are using stack memory which will be fast. We can re-visit it here, to use `alloca` intrinsics of C++ compiler here.